### PR TITLE
improve throughput by running smusher in parallel

### PR DESCRIPTION
**Pull request by the-architect**

### CODE REVIEW

File: lib/middleman-smusher/extension.rb [hunk 1]
```diff
@@ -17,21 +17,23 @@ def registered(app, options={})
...
-            
-              builder.say_status :smushed, file.sub(prefix, "")
-            end
+
+            builder.say_status :smushing, "%d files. This may take a moment." % files.size
+
+            files.map do |file|
+              Thread.new do
+                ::Smusher.optimize_image [file], options
+                builder.say_status :smushed, file.sub(prefix, '')
```
Comment (bhollis) [inline lib/middleman-smusher/extension.rb:53]: I think you'll need to create a `Mutex` and call `synchronize` around the `says_status` call, or else the output will get all messed up.

This implementation also may end up creating a _lot_ of threads - perhaps you could use a `Queue` to manage them?

**Comment by the-architect** [comment]
Yes, I'll do that. Though most of this should probably be handled by smusher itself sometime.

**Comment by the-architect** [comment]
@bhollis i think this is much better :)

**Comment by the-architect** [comment]
damn, getting strange behaviour from the queue.

**Comment by the-architect** [comment]
ok. this works :)

File: lib/middleman-smusher/extension.rb [hunk 2]
```diff
@@ -1,38 +1,65 @@
...
-                  
-            files.each do |file|
-              ::Smusher.optimize_image(
-                [file],
-                options
-              )
-            
-              builder.say_status :smushed, file.sub(prefix, "")
+
+            builder.say_status :smushing, "%d files. This may take a moment." % files.size
```
Comment (bhollis) [inline lib/middleman-smusher/extension.rb:24]: I don't think there's any need for this message.

File: lib/middleman-smusher/extension.rb [hunk 3]
```diff
@@ -1,38 +1,65 @@
 module Middleman
   module Smusher
+
+    DEFAULT_THREADS = 32
```
Comment (bhollis) [inline lib/middleman-smusher/extension.rb:4]: How did you choose this number? My guess is that anything more than the number of hardware threads on the system won't be productive. I think 4 or 8 would be a better guess (it's relatively hard to find the number of HW threads in a system-independent way from Ruby).

**Comment by the-architect** [comment]
Indeed, didn't know the queue class was THAT smart ;)

**Comment by bhollis** [comment]
Yeah, I don't know why they don't just call it BlockingQueue. Queue is a more generic thing to me.

**Comment by bhollis** [comment]
Hi @the-architect - any reason you closed this PR? It was basically done...

**Comment by the-architect** [comment]
I think i will take another approach to this. I am reinventing the wheel
here ;)

**Pull request closed without merging**